File: sdk/python/example/example_realtime_kline.py
import logging
import os
import time
from collections import defaultdict
from datetime import datetime
from typing import Dict

from pydantic import BaseModel

from kucoin_universal_sdk.api.client import DefaultClient
from kucoin_universal_sdk.generate.spot.spot_public.model_trade_event import TradeEvent
from kucoin_universal_sdk.model.client_option import ClientOptionBuilder
from kucoin_universal_sdk.model.constants import GLOBAL_API_ENDPOINT, GLOBAL_FUTURES_API_ENDPOINT, \
    GLOBAL_BROKER_API_ENDPOINT
from kucoin_universal_sdk.model.websocket_option import WebSocketClientOptionBuilder


class KLine(BaseModel):
    open: float
    high: float
    low: float
    close: float
    volume: float
    start_time: int
    end_time: int


kline_data: Dict[int, Dict[str, KLine]] = defaultdict(dict)

time_interval = 60


def on_trade_data(topic: str, subject: str, trade_event: TradeEvent):
    global kline_data

    symbol = trade_event.symbol
    price = float(trade_event.price)
    size = float(trade_event.size)
    timestamp = int(trade_event.time) // 1000000000

    period_start = timestamp // time_interval * time_interval
    period_end = period_start + time_interval

    if symbol not in kline_data[period_start]:
        kline = KLine(
            open=price,
            high=price,
            low=price,
            close=price,
            volume=size,
            start_time=period_start,
            end_time=period_end,
        )
        kline_data[period_start][symbol] = kline
    else:
        kline = kline_data[period_start][symbol]
        kline.high = max(kline.high, price)
        kline.low = min(kline.low, price)
        kline.close = price
        kline.volume += size

    print(f"KLine for {symbol} @ {datetime.fromtimestamp(period_start)}: {kline}")


def print_kline_data():
    for period_start in sorted(kline_data.keys()):
        print(f"\nTime Period: {datetime.fromtimestamp(period_start)}")
        for symbol, kline in kline_data[period_start].items():
            print(f"  Symbol: {symbol}")
            print(f"    Open: {kline.open}")
            print(f"    High: {kline.high}")
            print(f"    Low: {kline.low}")
            print(f"    Close: {kline.close}")
            print(f"    Volume: {kline.volume}")
            print(f"    Start Time: {datetime.fromtimestamp(kline.start_time)}")
            print(f"    End Time: {datetime.fromtimestamp(kline.end_time)}")


if __name__ == '__main__':
    logging.basicConfig(
        level=logging.INFO,
        format='%(asctime)s %(levelname)s - %(message)s',
        datefmt='%Y-%m-%d %H:%M:%S'
    )

    # Retrieve API secret information from environment variables
    key = os.getenv("API_KEY", "")
    secret = os.getenv("API_SECRET", "")
    passphrase = os.getenv("API_PASSPHRASE", "")

    # Set specific options, others will fall back to default values
    ws_client_option = (
        WebSocketClientOptionBuilder().build()
    )

    # Create a client using the specified options
    client_option = (
        ClientOptionBuilder()
        .set_key(key)
        .set_secret(secret)
        .set_passphrase(passphrase)
        .set_websocket_client_option(ws_client_option)
        .set_spot_endpoint(GLOBAL_API_ENDPOINT)
        .set_futures_endpoint(GLOBAL_FUTURES_API_ENDPOINT)
        .set_broker_endpoint(GLOBAL_BROKER_API_ENDPOINT)
        .build()
    )
    client = DefaultClient(client_option)

    spot_public = client.ws_service().new_spot_public_ws()
    spot_public.start()

    sub_id = spot_public.trade(["BTC-USDT", "ETH-USDT"], on_trade_data)

    time.sleep(180)

    spot_public.unsubscribe(sub_id)
    spot_public.stop()

    print_kline_data()

File: sdk/golang/example/example_realtime_kline.go
package main

import (
	"fmt"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/api"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/common/logger"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/generate/spot/spotpublic"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/types"
	"log"
	"os"
	"strconv"
	"sync"
	"time"
)

type KLine struct {
	Open      float64
	High      float64
	Low       float64
	Close     float64
	Volume    float64
	StartTime int64
	EndTime   int64
}

var (
	klineData    = make(map[int64]map[string]*KLine)
	klineMutex   = sync.Mutex{}
	timeInterval = int64(60)
)

func onTradeData(topic, subject string, tradeEvent *spotpublic.TradeEvent) error {
	klineMutex.Lock()
	defer klineMutex.Unlock()

	t, err := strconv.ParseInt(tradeEvent.Time, 10, 64)
	if err != nil {
		fmt.Println(err)
		return err
	}

	p, err := strconv.ParseFloat(tradeEvent.Price, 64)
	if err != nil {
		fmt.Println(err)
		return err
	}

	s, err := strconv.ParseFloat(tradeEvent.Size, 64)
	if err != nil {
		fmt.Println(err)
		return err
	}

	symbol := tradeEvent.Symbol
	price := p
	size := s
	timestamp := t / 1e9

	periodStart := (timestamp / timeInterval) * timeInterval
	periodEnd := periodStart + timeInterval

	if _, ok := klineData[periodStart]; !ok {
		klineData[periodStart] = make(map[string]*KLine)
	}

	if _, ok := klineData[periodStart][symbol]; !ok {
		klineData[periodStart][symbol] = &KLine{
			Open:      price,
			High:      price,
			Low:       price,
			Close:     price,
			Volume:    size,
			StartTime: periodStart,
			EndTime:   periodEnd,
		}
	} else {
		kline := klineData[periodStart][symbol]
		kline.High = max(kline.High, price)
		kline.Low = min(kline.Low, price)
		kline.Close = price
		kline.Volume += size
	}

	log.Printf("KLine for %s @ %s: %+v", symbol, time.Unix(periodStart, 0), klineData[periodStart][symbol])
	return nil
}

func printKlineData() {
	klineMutex.Lock()
	defer klineMutex.Unlock()

	for periodStart, symbols := range klineData {
		fmt.Printf("\nTime Period: %s\n", time.Unix(periodStart, 0))
		for symbol, kline := range symbols {
			fmt.Printf("  Symbol: %s\n", symbol)
			fmt.Printf("    Open: %.2f\n", kline.Open)
			fmt.Printf("    High: %.2f\n", kline.High)
			fmt.Printf("    Low: %.2f\n", kline.Low)
			fmt.Printf("    Close: %.2f\n", kline.Close)
			fmt.Printf("    Volume: %.2f\n", kline.Volume)
			fmt.Printf("    Start Time: %s\n", time.Unix(kline.StartTime, 0))
			fmt.Printf("    End Time: %s\n", time.Unix(kline.EndTime, 0))
		}
	}
}

func RealTimeKlineExample() {
	// Use the default logger or supply your custom logger
	defaultLogger := logger.NewDefaultLogger()
	logger.SetLogger(defaultLogger)

	// Retrieve API secret information from environment variables
	key := os.Getenv("API_KEY")
	secret := os.Getenv("API_SECRET")
	passphrase := os.Getenv("API_PASSPHRASE")

	// Set specific options, others will fall back to default values
	wsOption := types.NewWebSocketClientOptionBuilder().Build()

	// Create a client using the specified options
	option := types.NewClientOptionBuilder().
		WithKey(key).
		WithSecret(secret).
		WithPassphrase(passphrase).
		WithSpotEndpoint(types.GlobalApiEndpoint).
		WithFuturesEndpoint(types.GlobalFuturesApiEndpoint).
		WithBrokerEndpoint(types.GlobalBrokerApiEndpoint).
		WithWebSocketClientOption(wsOption).
		Build()
	client := api.NewClient(option)

	// Get the websocket service
	wsService := client.WsService()

	spotPublicWs := wsService.NewSpotPublicWS()
	err := spotPublicWs.Start()
	if err != nil {
		panic(err)
	}
	defer spotPublicWs.Stop()

	subId, err := spotPublicWs.Trade([]string{"BTC-USDT", "ETH-USDT"}, onTradeData)
	if err != nil {
		panic(err)
	}

	time.Sleep(3 * time.Minute)

	err = spotPublicWs.UnSubscribe(subId)
	if err != nil {
		panic(err)
	}

	printKlineData()

	log.Println("Program finished.")
}
